sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java

```diff
@@ -0,0 +1,315 @@
+          FieldFilter.Operator.LESS_THAN_OR_EQUAL,
+          FieldFilter.Operator.GREATER_THAN,
+          FieldFilter.Operator.GREATER_THAN_OR_EQUAL,
+          FieldFilter.Operator.NOT_EQUAL,
+          FieldFilter.Operator.NOT_IN);
+  private static final ImmutableSet<UnaryFilter.Operator> INEQUALITY_UNARY_FILTER_OPS =
+      ImmutableSet.of(UnaryFilter.Operator.IS_NOT_NAN, UnaryFilter.Operator.IS_NOT_NULL);
+
+  private static final String UNQUOTED_NAME_REGEX_STRING = "([a-zA-Z_][a-zA-Z_0-9]*)";
+  private static final String QUOTED_NAME_REGEX_STRING = "(`(?:[^`\\\\]|(?:\\\\.))+`)";
+  // After each segment follows a dot and more characters, or the end of the string.
+  private static final Pattern FIELD_PATH_SEGMENT_REGEX =
+      Pattern.compile(
+          String.format("(?:%s|%s)(\\..+|$)", UNQUOTED_NAME_REGEX_STRING, QUOTED_NAME_REGEX_STRING),
+          Pattern.DOTALL);
+
+  static List<Order> getImplicitOrderBy(StructuredQuery query) {
+    List<String> expectedImplicitOrders = new ArrayList<>();
+    if (query.hasWhere()) {
+      fillInequalityFields(query.getWhere(), expectedImplicitOrders);
+    }
+    if (!expectedImplicitOrders.contains("__name__")) {
+      expectedImplicitOrders.add("__name__");
+    }
+    for (Order order : query.getOrderByList()) {
+      String orderField = order.getField().getFieldPath();
+      expectedImplicitOrders.remove(orderField);
+    }
+
+    List<Order> additionalOrders = new ArrayList<>();
+    if (!expectedImplicitOrders.isEmpty()) {
+      Direction lastDirection =
+          query.getOrderByCount() == 0
+              ? Direction.ASCENDING
+              : query.getOrderByList().get(query.getOrderByCount() - 1).getDirection();
+
+      for (String field : expectedImplicitOrders) {
+        additionalOrders.add(
+            Order.newBuilder()
+                .setDirection(lastDirection)
+                .setField(FieldReference.newBuilder().setFieldPath(field).build())
+                .build());
+      }
+    }
+
+    return additionalOrders;
+  }
+
+  private static void fillInequalityFields(Filter filter, List<String> result) {
+    switch (filter.getFilterTypeCase()) {
+      case FIELD_FILTER:
+        if (INEQUALITY_FIELD_FILTER_OPS.contains(filter.getFieldFilter().getOp())) {
+          String fieldPath = filter.getFieldFilter().getField().getFieldPath();
+          if (!result.contains(fieldPath)) {
+            result.add(fieldPath);
+          }
+        }
+        break;
+      case COMPOSITE_FILTER:
+        filter.getCompositeFilter().getFiltersList().forEach(f -> fillInequalityFields(f, result));
+        break;
+      case UNARY_FILTER:
+        if (INEQUALITY_UNARY_FILTER_OPS.contains(filter.getUnaryFilter().getOp())) {
+          String fieldPath = filter.getUnaryFilter().getField().getFieldPath();
+          if (!result.contains(fieldPath)) {
+            result.add(fieldPath);
+          }
+        }
+        break;
+      default:
+        break;
+    }
+  }
```

Comment:
@yixiaoshen was saying that the filters in a composite filter would already be sorted. But since StructuredQuery is a raw proto, I'm guessing this would only be the case on the backend?
If not I'll add sorting to the composite filter case, and the default case should probably throw a NotImplemented instead of a break.

Reply:
> that the filters in a composite filter would already be sorted.

The filter order in a composite filter is arbitrary, and picked by the user.

> If not I'll add sorting to the composite filter case

Composite filters can be nested, the sorting needs to be done for all field paths that appear at any point in the query filter "tree". I think it's also reasonable to not sort and only ensure that the results contains only one *canonical* field paths, otherwise fail with a clear message that **this** version of the beam connector does not support query with multiple inequalities (this is punting the issue for later, and hoping we will have a better mechanism for resumption by then).

Reply:
I went ahead and implemented a class modeled after [FieldPath](https://github.com/googleapis/java-firestore/blob/main/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FieldPath.java) with the extra escaping logic but without support for non-Where field paths. Just adding a compareTo there was easy, so the sorting is fully implemented now.